cdap-app-fabric/src/main/java/io/cdap/cdap/app/preview/DefaultPreviewRunnerManager.java

```diff
@@ -124,21 +125,32 @@ protected void startUp() throws Exception {
 
     // Create and start the preview poller services.
     for (int i = 0; i < maxConcurrentPreviews; i++) {
-      String pollerInfo = UUID.randomUUID().toString();
+      PreviewRunnerService pollerService = previewRunnerServiceFactory.create(runner);
 
-      PreviewRunnerService pollerService = new PreviewRunnerService(
-        previewCConf, previewInjector.getInstance(PreviewRunner.class),
-        previewRequestFetcherFactory.create(Bytes.toBytes(pollerInfo)));
+      pollerService.addListener(new ServiceListenerAdapter() {
+        @Override
+        public void terminated(State from) {
+          previewRunnerServices.remove(pollerService);
+          if (previewRunnerServices.isEmpty()) {
+            try {
+              stopAndWait();
```

Comment:
I don't think you need to wait. Just call `stop()` would be ok.

Reply:
done.